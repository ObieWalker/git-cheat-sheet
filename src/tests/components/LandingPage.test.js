import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import connect, { LandingPage } from '../../components/LandingPage';
let mountedLogin;
let props;
const locations = [];
const history = {
  push(location) {
    locations.push(location);
  }
};
const command = "$ git checkout to there"

const userData = { isAuthenticated: false, user: {} };
const getAllCheats = jest.fn(() => Promise.resolve());
const addCategory = jest.fn(() => Promise.resolve());
const deleteCategory = jest.fn(() => Promise.resolve());
const deleteCheat = jest.fn(() => Promise.resolve());
const editCategory = jest.fn(() => Promise.resolve());
const addCheat = jest.fn(() => Promise.resolve());
const logOut = jest.fn(() => Promise.resolve());
const mockStore = configureStore();
let wrapper;
const cheats = {
  name: "A name",
  _id: 1211
}
const category = {
  name: 'random'
}

const getComponent = () => {
  if (!mountedLogin) {
    props = {
      userData,
      getAllCheats,
      addCategory,
      deleteCategory,
      deleteCheat,
      editCategory,
      addCheat, 
      logOut,
      history
    };
    history.push = jest.fn();
    mountedLogin = shallow(<LandingPage {...props} />);
  }
  return mountedLogin;
};

describe('Landing page Component', () => {
  beforeEach(() => {});
  wrapper = getComponent();

  it('renders component successfully', () => {
    wrapper = getComponent();
    expect(wrapper).toMatchSnapshot();
  });
  describe('Check functions are called', () => {
    it('should have a method handleLogOut', () => {
      expect(wrapper.instance().handleLogOut).toBeDefined();
    });
    it('should have a method that addCategory', () => {
      expect(wrapper.instance().addCategory).toBeDefined();
    });
    it('should have a method that addCheat', () => {
      expect(wrapper.instance().addCheat).toBeDefined();
    });
    it('should have a method that editCategory', () => {
      expect(wrapper.instance().editCategory).toBeDefined();
    });
    it('should have a method that deleteCategory', () => {
      expect(wrapper.instance().deleteCategory).toBeDefined();
    });
    it('should have a method that deleteCheat', () => {
      expect(wrapper.instance().deleteCheat).toBeDefined();
    });
    it('should have a method that handleHide', () => {
      expect(wrapper.instance().handleHide).toBeDefined();
    });
    it('should have a method that cheatModalOnHide', () => {
      expect(wrapper.instance().cheatModalOnHide).toBeDefined();
    });
    it('should have a method that cheatModalShow', () => {
      expect(wrapper.instance().cheatModalShow).toBeDefined();
    });
    it('should have a method that modalShow', () => {
      expect(wrapper.instance().modalShow).toBeDefined();
    });
    it('should have a method that updateDetails', () => {
      expect(wrapper.instance().updateDetails).toBeDefined();
    });
    it('should have a method that updateSearch', () => {
      expect(wrapper.instance().updateSearch).toBeDefined();
    });
    it('should have a method that onCopy', () => {
      expect(wrapper.instance().onCopy).toBeDefined();
    });
    it('should have a method that editformIsValid', () => {
      expect(wrapper.instance().editformIsValid).toBeDefined();
    });
    it('should have a method that addCategoryformIsValid', () => {
      expect(wrapper.instance().addCategoryformIsValid).toBeDefined();
    });
    it('should have a method that addCheatformIsValid', () => {
      expect(wrapper.instance().addCheatformIsValid).toBeDefined();
    });
  })

  describe('Checks state change', () => {
    it('should have a method handleHide', () => {
      wrapper = getComponent();
      wrapper.instance().handleHide();
      expect(wrapper.state().show).toEqual(false);
      expect(wrapper.state().categoryTitle).toEqual('');
      expect(wrapper.state().description).toEqual('');
      expect(wrapper.state().command).toEqual('');
      expect(wrapper.state().category).toEqual({});
      expect(wrapper.state().errors).toEqual({});
    });
    it('should have a method cheatModalOnHide', () => {
      wrapper = getComponent();
      wrapper.instance().cheatModalOnHide();
      expect(wrapper.state().cheatModalShow).toEqual(false);
      expect(wrapper.state().description).toEqual('');
      expect(wrapper.state().command).toEqual('');
      expect(wrapper.state().keywords).toEqual('');
      expect(wrapper.state().errors).toEqual({});
    });
    it('should have a method cheatModalShow', () => {
      wrapper = getComponent();
      wrapper.instance().cheatModalShow(cheats);
      expect(wrapper.state().cheatModalShow).toEqual(true);
      expect(wrapper.state().categoryTitle).toEqual('A name');
      expect(wrapper.state().category.id).toEqual(1211);
    });
    it('should have a method modalShow', () => {
      wrapper = getComponent();
      wrapper.instance().modalShow(category);
      expect(wrapper.state().category).toEqual(category);
      expect(wrapper.state().show).toEqual(true);
    });
    it('should click on on show', () => {
      wrapper = getComponent();
      wrapper
      .find('.btn-primary')
      .simulate('click');
      expect(wrapper.state().show).toEqual(true);
    });  
    it('should have a method handleLogOut', () => {
      wrapper = getComponent();
      const e = { preventDefault: () => undefined };
      e.preventDefault = jest.fn();
      wrapper.instance().handleLogOut(e);
      expect(e.preventDefault).toHaveBeenCalled();
    });
    it('should have a method addCategory', () => {
      wrapper = getComponent();
      wrapper.setState({
        categoryTitle: 'Center Place',
        userData: {
          user: {
            id: 1222
          }
        }
      });
      const e = { preventDefault: () => undefined };
      e.preventDefault = jest.fn();
      wrapper.instance().addCategory(e);
      expect(wrapper.state().show).toEqual(true);
    });
    it('should have a method updateSearch', () => {
      wrapper = getComponent();
      wrapper.setState({
        search: 'Cent',
      });
      const e = { 
        preventDefault: () => undefined,
        target: {
          name: 'Center Place'
        }
       };
      e.preventDefault = jest.fn();
      wrapper.instance().updateSearch(e);
      expect(wrapper.state().search).toEqual('Cent');
    });
    it('should have a method updateDetails', () => {
      wrapper = getComponent();
      wrapper.setState({
        categoryTitle: 'Center Place',
      });
      const e = { 
        preventDefault: () => undefined,
        target: {
          name: 'Center Place'
        }
       };
      e.preventDefault = jest.fn();
      wrapper.instance().updateDetails(e);
      expect(wrapper.state().categoryTitle).toEqual('Center Place');
    });
    it('should have a method addCheat', () => {
      wrapper = getComponent();
      wrapper.setState({
        command: 'Center Place',
        description: 'random description',
        keywords: 'blank',
      });
      wrapper.instance().addCheat();
      expect(wrapper.state().cheatModalShow).toEqual(true);
    });
    it('should have a method editCategory', () => {
      wrapper = getComponent();
      wrapper.setState({
        categoryTitle: 'Center Place',
        userData: {
          user: {
            id: 1222
          }
        }
      });
      wrapper.instance().editCategory();
      expect(wrapper.state().show).toEqual(false);
    });
    it('should have a method deleteCategory', () => {
      wrapper = getComponent();
      wrapper.setState({
        categoryTitle: 'Center Place'
      });
      wrapper.instance().deleteCategory();
    });
    it('should have a method deleteCheat', () => {
      wrapper = getComponent();
      wrapper.instance().deleteCheat();
    });
    it('should have a method editformIsValid', () => {
      wrapper = getComponent();
      wrapper.setState({
        categoryTitle: ''
      });
      wrapper.instance().editformIsValid();
      expect(wrapper.state().errors).toEqual({"categoryTitle": "There must be valid change to the category title."});
    });
    it('should have a method addCategoryformIsValid', () => {
      wrapper = getComponent();
      wrapper.setState({
        categoryTitle: ''
      });
      wrapper.instance().addCategoryformIsValid();
      expect(wrapper.state().errors).toEqual({"categoryTitle": "The Category title cannot be empty."});
    });
    it('should have a method addCheatformIsValid', () => {
      wrapper = getComponent();
      wrapper.setState({
        description: ''
      });
      wrapper.instance().addCheatformIsValid();
      expect(wrapper.state().errors).toEqual({
        "description": "There must be a description.",
        "command": "Command cannot be empty.",
      });
    });
  })
})

describe('Connected Component', () => {
  describe('Connected mountedLanding page', () => {
    it('component successfully rendered', () => {
      const store = mockStore({
        userData,
        getAllCheats,
        addCategory,
        deleteCategory,
        deleteCheat,
        editCategory,
        addCheat, 
        logOut
      });
      wrapper = shallow(<connect 
        store={store} 
        history={history}
        getAllCheats={getAllCheats}
        addCategory={addCategory}
        deleteCategory={deleteCategory}
        deleteCheat={deleteCheat}
        editCategory={editCategory}
        addCheat={addCheat}
        logOut={logOut}
        />);
      expect(wrapper.length).toBe(1);
    });
  });
});
