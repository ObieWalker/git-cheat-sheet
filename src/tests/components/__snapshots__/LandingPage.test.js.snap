// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing page Component renders component successfully 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LandingPage
    addCategory={[MockFunction]}
    addCheat={[MockFunction]}
    deleteCategory={[MockFunction]}
    deleteCheat={[MockFunction]}
    editCategory={[MockFunction]}
    getAllCheats={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Promise {},
          },
        ],
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    logOut={[MockFunction]}
    userData={
      Object {
        "isAuthenticated": false,
        "user": Object {},
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="waves-effect waves-dark pull-right hoverable dark-green btn btn-danger"
          id="add-grocery"
          onClick={[Function]}
          type="submit"
        >
           Logout
        </button>,
        <div>
          <SearchBar
            updateSearch={[Function]}
          />
          <button
            className="waves-effect waves-dark hoverable dark-green btn btn-primary"
            id="add-grocery"
            onClick={[Function]}
            type="submit"
          >
            Create new Category
          </button>
        </div>,
        <br />,
        <div
          className="container"
        >
          <Connect(Categories)
            cheatModalShow={[Function]}
            deleteCategory={[Function]}
            deleteCheat={[Function]}
            onCopy={[Function]}
            onShow={[Function]}
            search=""
          />
        </div>,
        <CategoryModal
          addCategory={[Function]}
          category={Object {}}
          editCategory={[Function]}
          errors={Object {}}
          onHide={[Function]}
          show={false}
          updateDetails={[Function]}
        />,
        <CheatModal
          addCheat={[Function]}
          category={Object {}}
          categoryTitle=""
          cheatModalShow={false}
          errors={Object {}}
          onHide={[Function]}
          show={false}
          updateDetails={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Logout",
          "className": "waves-effect waves-dark pull-right hoverable dark-green btn btn-danger",
          "id": "add-grocery",
          "onClick": [Function],
          "type": "submit",
        },
        "ref": null,
        "rendered": " Logout",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <SearchBar
              updateSearch={[Function]}
            />,
            <button
              className="waves-effect waves-dark hoverable dark-green btn btn-primary"
              id="add-grocery"
              onClick={[Function]}
              type="submit"
            >
              Create new Category
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "updateSearch": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create new Category",
              "className": "waves-effect waves-dark hoverable dark-green btn btn-primary",
              "id": "add-grocery",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "Create new Category",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Connect(Categories)
            cheatModalShow={[Function]}
            deleteCategory={[Function]}
            deleteCheat={[Function]}
            onCopy={[Function]}
            onShow={[Function]}
            search=""
          />,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cheatModalShow": [Function],
            "deleteCategory": [Function],
            "deleteCheat": [Function],
            "onCopy": [Function],
            "onShow": [Function],
            "search": "",
            "userId": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "addCategory": [Function],
          "category": Object {},
          "editCategory": [Function],
          "errors": Object {},
          "onHide": [Function],
          "show": false,
          "updateDetails": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "addCheat": [Function],
          "category": Object {},
          "categoryTitle": "",
          "cheatModalShow": false,
          "errors": Object {},
          "onHide": [Function],
          "show": false,
          "updateDetails": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="waves-effect waves-dark pull-right hoverable dark-green btn btn-danger"
            id="add-grocery"
            onClick={[Function]}
            type="submit"
          >
             Logout
          </button>,
          <div>
            <SearchBar
              updateSearch={[Function]}
            />
            <button
              className="waves-effect waves-dark hoverable dark-green btn btn-primary"
              id="add-grocery"
              onClick={[Function]}
              type="submit"
            >
              Create new Category
            </button>
          </div>,
          <br />,
          <div
            className="container"
          >
            <Connect(Categories)
              cheatModalShow={[Function]}
              deleteCategory={[Function]}
              deleteCheat={[Function]}
              onCopy={[Function]}
              onShow={[Function]}
              search=""
            />
          </div>,
          <CategoryModal
            addCategory={[Function]}
            category={Object {}}
            editCategory={[Function]}
            errors={Object {}}
            onHide={[Function]}
            show={false}
            updateDetails={[Function]}
          />,
          <CheatModal
            addCheat={[Function]}
            category={Object {}}
            categoryTitle=""
            cheatModalShow={false}
            errors={Object {}}
            onHide={[Function]}
            show={false}
            updateDetails={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Logout",
            "className": "waves-effect waves-dark pull-right hoverable dark-green btn btn-danger",
            "id": "add-grocery",
            "onClick": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": " Logout",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SearchBar
                updateSearch={[Function]}
              />,
              <button
                className="waves-effect waves-dark hoverable dark-green btn btn-primary"
                id="add-grocery"
                onClick={[Function]}
                type="submit"
              >
                Create new Category
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "updateSearch": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Create new Category",
                "className": "waves-effect waves-dark hoverable dark-green btn btn-primary",
                "id": "add-grocery",
                "onClick": [Function],
                "type": "submit",
              },
              "ref": null,
              "rendered": "Create new Category",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Connect(Categories)
              cheatModalShow={[Function]}
              deleteCategory={[Function]}
              deleteCheat={[Function]}
              onCopy={[Function]}
              onShow={[Function]}
              search=""
            />,
            "className": "container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "cheatModalShow": [Function],
              "deleteCategory": [Function],
              "deleteCheat": [Function],
              "onCopy": [Function],
              "onShow": [Function],
              "search": "",
              "userId": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addCategory": [Function],
            "category": Object {},
            "editCategory": [Function],
            "errors": Object {},
            "onHide": [Function],
            "show": false,
            "updateDetails": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addCheat": [Function],
            "category": Object {},
            "categoryTitle": "",
            "cheatModalShow": false,
            "errors": Object {},
            "onHide": [Function],
            "show": false,
            "updateDetails": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
